pkg/controller/workload/jobframework/helpers.go

```diff
@@ -166,7 +166,11 @@ func StopJob(ctx context.Context, client client.Client, record record.EventRecor
 	}
 
 	if wl != nil {
-		if err := job.RestoreNodeAffinity([]map[string]string{wl.Spec.PodSets[0].Template.Spec.NodeSelector}); err != nil {
+		nodeSelectors := []map[string]string{}
+		for _, podSet := range wl.Spec.PodSets {
+			nodeSelectors = append(nodeSelectors, podSet.Template.Spec.NodeSelector)
+		}
+		if err := job.RestoreNodeAffinity(nodeSelectors); err != nil {
```

Comment:
to avoid the allocation, I would just pass the list of PodSets to `RestoreNodeAffinity`